### DISCUSSION TITLE
Wave Generator -> I2S -> Format convertor (16,32) -> DAC ES9018k2m

### DISCUSSION THREAD

**Discussion by TangerineDreamer** [pschatzmann/arduino-audio-tools#1484] [Q&A] [answered]
Hello,
have a hard time trying to get a good sound with that DAC,but always have a pure wave mixed
with a steam machine-like,here's my code:

```
#include "AudioTools.h"

AudioInfo info(32000,2,16);
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
I2SStream out;
NumberFormatConverterStream fc(out); 
StreamCopy copier(out, sound);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Setup sine wave
  sineWave.begin(info, 1000.00);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);

  //config.i2s_format = I2S_STD_FORMAT;
  //config.is_master = false;
  config.pin_bck = 14;//11
config.pin_ws = 15;//10
config.pin_data = 22;//12
  
  
//config.bits_per_sample = 32; // we coult do this explicitly
   
  out.begin(config);
  fc.begin(16, 32);
  

  
  Serial.println("started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
```
That don't really change if i put fc.begin() before out.begin() 
Did i do something wrong?

**Comment by pschatzmann** [comment, marked as answer]
What does the log say ? 
I think you just forgot to configure I2S as 32 bits, so it is still on 16 bits but you send it 32 bit data!

**Reply by TangerineDreamer** [reply]
So i changed the i2s info to 32bits,the steam-like sound diseappeared,but a problem appears with the AudioLogger,
when the output is Info,the sound is not pured has interruptions all on the way,when the output is Warning,the sound is pure
but also interrupted,here's my new sketch and audiologger infos: 

```
#include "AudioTools.h"

AudioInfo sound_16(32000,2,16);
AudioInfo i2s_32(32000,2,32);
SineWaveGenerator<int16_t> sineWave(32000);                // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> sound(sineWave);             // Stream generated from sine wave
I2SStream out;
NumberFormatConverterStream fc(out); 
StreamCopy copier(out, sound);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  while(!Serial);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // Setup sine wave
  sineWave.begin(sound_16, 1000.00);

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.copyFrom(i2s_32);

  //config.i2s_format = I2S_STD_FORMAT;
  //config.is_master = true;
  config.pin_bck = 14;//11
config.pin_ws = 15;//10
config.pin_data = 22;//12
  
  
//config.bits_per_sample = 32; // we coult do this explicitly
   
  out.begin(config);
  fc.begin(16, 32);
  

  
  Serial.println("started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
```

AUDIOLOGGER:

[I] SoundGenerator.h : 167 - SineWaveGenerator::begin(channels=2, sample_rate=32000, frequency=1000.00)
[I] SoundGenerator.h : 151 - bool audio_tools::SineWaveGenerator<T>::begin() [with T = short int]
[I] SoundGenerator.h : 44 - bits_per_sample: 16
[I] SoundGenerator.h : 45 - channels: 2
[I] SoundGenerator.h : 46 - sample_rate: 32000
[I] Buffers.h : 372 - resize: 4
[I] SoundGenerator.h : 194 - setFrequency: 1000.00
[I] SoundGenerator.h : 195 - active: true
starting I2S...
[I] AudioTypes.h : 124 - sample_rate: 32000
[I] AudioTypes.h : 125 - channels: 2
[I] AudioTypes.h : 126 - bits_per_sample: 32
[I] AudioTypes.h : 124 - sample_rate: 32000
[I] AudioTypes.h : 125 - channels: 2
[I] AudioTypes.h : 126 - bits_per_sample: 32
[I] I2SConfigESP32.h : 80 - rx/tx mode: TX_MODE
[I] I2SConfigESP32.h : 81 - port_no: 0
[I] I2SConfigESP32.h : 82 - is_master: Master
[I] I2SConfigESP32.h : 83 - sample rate: 32000
[I] I2SConfigESP32.h : 84 - bits per sample: 32
[I] I2SConfigESP32.h : 85 - number of channels: 2
[I] I2SConfigESP32.h : 86 - signal_type: Digital
[I] I2SConfigESP32.h : 88 - i2s_format: I2S_STD_FORMAT
[I] I2SConfigESP32.h : 90 - auto_clear: true
[I] I2SConfigESP32.h : 92 - use_apll: true
[I] I2SConfigESP32.h : 97 - buffer_count:6
[I] I2SConfigESP32.h : 98 - buffer_size:512
[I] I2SConfigESP32.h : 103 - pin_bck: 14
[I] I2SConfigESP32.h : 105 - pin_ws: 15
[I] I2SConfigESP32.h : 107 - pin_data: 22
[I] AudioStreamsConverter.h : 419 - begin 16 -> 32 bits
started...
[I] StreamCopy.h : 149 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops..........and so on

**Comment by pschatzmann** [comment]
What do you mean with pure but interrupted ?

I can't reproduce your problem and for me it is working perfectly with log level Warning!
Maybe you need to update the library.
